src/main/java/sk/uniba/fmph/dcs/game_board/VariableBuilding.java

```diff
@@ -0,0 +1,45 @@
+package sk.uniba.fmph.dcs.game_board;
+
+import sk.uniba.fmph.dcs.stone_age.Effect;
+
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.OptionalInt;
+
+public final class VariableBuilding implements Building {
+    private final int numberOfResources;
+    private final int numberOfResourcesTypes;
+
+    public VariableBuilding(final int numberOfResources, final int numberOfResourcesTypes) {
+        this.numberOfResources = numberOfResources;
+        this.numberOfResourcesTypes = numberOfResourcesTypes;
+        if (numberOfResources <= 0 || numberOfResourcesTypes <= 0) {
+            throw new IllegalArgumentException("numberOfResources/numberOfResourceTypes must be greater than 0");
+        }
+        if (numberOfResourcesTypes > numberOfResources) {
+            throw new IllegalArgumentException("numberOfResourceTypes must be less or equal than numberOfResources");
+        }
+    }
+
+    @Override
+    public OptionalInt build(final Collection<Effect> resources) {
+        if (resources.size() != numberOfResources) {
+            return OptionalInt.empty();
+        }
+        int sum = 0;
+        ArrayList<Effect> wasResourceType = new ArrayList<>();
+        for (Effect resource : resources) {
+            if (!resource.isResource()) {
+                return OptionalInt.empty();
+            }
+            if (!wasResourceType.contains(resource)) {
+                wasResourceType.add(resource);
+            }
```

Comment:
```suggestion
        Set<Effect> wasResourceType = new HashSet<>();
        for (Effect resource : resources) {
            if (!resource.isResource()) {
                return OptionalInt.empty();
            }
            wasResourceType.add(resource);
```
Teoreticky je tu lepšie použiť `HashSet` namiesto `ArrayList`, lebo zistenie, či prvok patrí do `HashSet` je O(1), oproti `ArrayList`, kde je to O(n). Na druhej strane predpokladáme, že tento `ArrayList` bude veľmi malý, preto na tomto až tak nezáleží. Podľa internej implementácie hashovania je možno aj `ArrayList` pre takto malé polia rýchlejší, ťažko povedať.

Túto zmenu nechám na teba, či ju tam pridáš alebo nie, je to v podstate jedno.

Reply:
Vyzera to prehladnejsie ked je to Hashset.

Reply:
Okej